gtk4/src/subclass/im_context.rs

```diff
@@ -0,0 +1,512 @@
+    ptr: *mut ffi::GtkIMContext,
+    stringptr: *const libc::c_char,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let string: Borrowed<GString> = from_glib_borrow(stringptr);
+
+    imp.commit(wrap.unsafe_cast_ref(), string.as_str())
+}
+
+unsafe extern "C" fn im_context_delete_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    offset: i32,
+    n_chars: i32,
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.delete_surrounding(wrap.unsafe_cast_ref(), offset, n_chars)
+        .to_glib()
+}
+
+unsafe extern "C" fn im_context_filter_keypress<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    eventptr: *mut gdk::ffi::GdkEvent,
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+    let event: Borrowed<gdk::Event> = from_glib_borrow(eventptr);
+    imp.filter_keypress(wrap.unsafe_cast_ref(), &event)
+        .to_glib()
+}
+
+unsafe extern "C" fn im_context_focus_in<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.focus_in(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_focus_out<T: IMContextImpl>(ptr: *mut ffi::GtkIMContext) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    imp.focus_out(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn im_context_get_preedit_string<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    text_ptr: *mut *mut libc::c_char,
+    attrs_ptr: *mut *mut pango::ffi::PangoAttrList,
+    cursor_index_ptr: *mut libc::c_int,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    let (text, attrs, cursor_idx) = imp.get_preedit_string(wrap.unsafe_cast_ref());
+
+    *text_ptr = text.to_glib_none().0;
+    *cursor_index_ptr = cursor_idx;
+    *attrs_ptr = attrs.to_glib_none().0;
+}
+
+unsafe extern "C" fn im_context_get_surrounding<T: IMContextImpl>(
+    ptr: *mut ffi::GtkIMContext,
+    text_ptr: *mut *mut libc::c_char,
+    cursor_index_ptr: *mut libc::c_int,
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<IMContext> = from_glib_borrow(ptr);
+
+    if let Some((text, cursor_idx)) = imp.get_surrounding(wrap.unsafe_cast_ref()) {
+        *text_ptr = text.to_glib_none().0;
+        *cursor_index_ptr = cursor_idx;
+        return true.to_glib();
+    }
+    false.to_glib()
```

Comment:
Maybe to be friendly you want to set `text_ptr` and `cursor_index_ptr` to `NULL` / `0` in this case